src/main/python/run_trecqa.py

```diff
@@ -5,219 +5,294 @@
+                f.write('{} 0 {} {}\n'.format(qid, answer[0], answer[2]))
 
+# evaluate by TREC QA patterns
+def eval_by_pattern(qid, candidates, pattern, out):
+    seen = set([])
 
-def get_precision(actual, predicted, k):
-  total_count = 0.0
-  hits = 0.0
+    for i, cand in enumerate(candidates):
+        this_candidate = " ".join(cand[0])
 
-  for qid in predicted.keys():
-    run_labels = predicted.copy()
+        hash_value = hashlib.md5(this_candidate.encode()).hexdigest()
+        answer_id = "QA{}.{}".format(qid, hash_value)
 
-    if len(predicted[qid]) > k:
-      run_labels = run_labels[qid][:k]
-    else:
-      run_labels = run_labels[qid]
+        if answer_id in seen:
+            continue
+        seen.add(answer_id)
 
-    for p in run_labels:
-      total_count += 1
+        out.write('{} Q0 {} {} {} TrecQA-pattern\n'.format(qid, answer_id, i+1, 1.0/(i + 1)))
 
-      if not actual[qid]:
-        continue
+# evaluate the run file at different depths
+def evaluate_at_k(fname, eval_depth):
+    for k in eval_depth:
+        with open(fname) as f,  open("{}.k{}.txt".format(fname, k), 'w') as output_file:
+            for line in f:
+                det = line.strip().split()
+                rank = int(det[3])
 
-      if "unjudged" in p:
-        hits += 0
-      else:
-        hits += actual[qid][p]
+                if rank <= int(k):
+                    output_file.write(line)
 
-  if total_count > 0:
-    return hits / total_count
-  else:
-    return 0
 
+def create_qrel_pattern(all_sentences, pattern, qrels, qid):
+    seen = set([])
+    for i, cand in enumerate(all_sentences):
+        this_candidate = cand.lower().split("\t")[0]
+        result = re.findall(r'{}'.format(pattern.lower()), this_candidate)
+        hash_value = hashlib.md5(this_candidate.encode()).hexdigest()
+        answer_id = "QA{}.{}".format(qid, hash_value)
+        if answer_id in seen:
+            continue
+        seen.add(answer_id)
+
+        if result:
+            qrels.write('{} 0 {} {}\n'.format(qid, answer_id, 1))
+        else:
+            qrels.write('{} 0 {} {}\n'.format(qid, answer_id, 0))
+
+# sentences with at least one nonstop question words
+def clean_sentences(question, all_sentences, stop_words):
+    question_words = set(question.strip().split())
+    non_stop_question_words = question_words - stop_words
+    cleaned_sentences = []
+
+    for sent in all_sentences:
+        this_candidate = sent.lower().split("\t")[0]
+        this_candidate_tokens = set(this_candidate.split())
+
+        word_overlap = this_candidate_tokens.intersection(non_stop_question_words)
+
+        if len(word_overlap) > 0:
+            cleaned_sentences.append(sent)
+
+    return cleaned_sentences
+
+# load the stopwords from the resource file
+def get_stopwords():
+    stopwords = set([])
+    with open("src/main/resources/io/anserini/qa/english-stoplist.txt") as stop:
+        for line in stop:
+            if "#" not in line:
```

Comment:
`# not in line` or `not line.startswith("#")`? There's a subtle difference and even if it doesn't matter in this case (which I suspect it doesn't), the intent should be made clearer.